### ISSUE TITLE
break build systems and dependency `type`s down into more granular resource dependencies

### ISSUE THREAD

**Issue by cosmicexplorer** [spack/spack#20523]
**Probably blocking on #20530 to resolve this, but the type of solution we get to #20530 might come from the discussion in this issue.**

This is a similar attempt to #10411, but with a broader scope. We attempt to explicitly represent interaction with build systems by introducing the "resource dependencies" concept instead of defining a dependency by build step names alone.

**Unlike #20256, *any* extensions of the spec syntax are out of scope for this issue.** Once this is done, we would continue to allow build, link, run, and test dependencies, but internally they would each be implemented as an *umbrella virtual dependency*, covering each *underlying resource dependency*, **for no change in functionality.**

### Rationale

1. We have long desired to make a package work in more than 1 type of Spack "build system" at a time, e.g. if a package switches from using autotools to cmake. See #12941 for context.
    - https://github.com/spack/spack/pull/12941#issuecomment-538870806 describes the current consensus on use cases, and tries to lay out one possible solution.
    - Valiant attempts to shoehorn the existing inheritance model such as #10411 have found difficulty convincing others to merge them.

2. Separately from build systems, it also seems spack may be conflating several concepts into the dependency `type`s that can be declared in package.py files. An excerpt like:  
> X RPATHs Y (EDIT) and/or Y is added to library paths for X at run time
> X adds -I for Y's include directories when building

from this comment https://github.com/spack/spack/issues/3768#issuecomment-293046651 (quoted immediately above) seems to point out several very specific dependency relationships which **don't** seem to fit into the current model, **(where a build "type" determines multiple very specific things, despite having the same name as a build "step")**.

There was a discussion just now with @tgamblin and @scheibelp where it was agreed that the definition of a spack "test" dependency seemed to be the least obvious of this set of words that mean two very different things:
> “test”: the dependency will be added to the PATH and PYTHONPATH at build-time. The only difference between “build” and “test” is that test dependencies are only built if the user requests unit tests with spack install --test.

3. We already define our dependency types in terms of other much more granular information here: https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=dependency%20type#dependency-types. While this is at least clearly exposed in our docs, why not make an explicit API for that much more granular information to allow for dependencies like in https://github.com/spack/spack/issues/3768#issuecomment-293046651, especially when `spack python` exists?

Even with just two dependencies at once like at the bottom of that comment, it seems to become very difficult to predict the result when Spack evaluates the spec. Furthermore, it seems to me that having to go back and exhaustively post-hoc categorize all of the *intended* changes as is done in that comment is *far* more effort than writing this code with an API that explicitly accesses and provides these very specific resources.

That precise argument also seems to apply to the current situation of monolithic, non-composable build system abstractions. I believe there may be a solution to both of the perceived issues with monolithic build systems and monolithic dependency types at once.

### Description

This is a general approach I'm thinking might address this:

#### Require Explicit Dependencies on `.libs` and `.headers`, and Validate Them

This is a paradigm that [pants](https://github.com/pantsbuild/pants), another build tool, has been moving to over the past few years for defining packages ("targets" in pants parlance). The site has a great blurb on what this means (at https://www.pantsbuild.org/docs/target-api-concepts#a-field-driven-api):

> Pants plugins do not care about specific target types; they only care that the target type has the right combination of field types that the plugin needs to operate.
>
> For example, the Python autoformatter Black does not actually care whether you have a python_library, python_tests, or custom_target target; all that it cares about is that your target type has the field PythonSources.

*I ranted and raved about this to @becker33 a while ago: pants implements this via this very basic "UnionRule" concept I made up: https://github.com/pantsbuild/pants/pull/7116. I don't think any of the pants rule DSL is relevant to this, but I can explain the mechanics of it further if necessary.*

And the part I think is particularly relevant here is that **`Field` has its own entire class hierarchy which decouples validating outputs from consuming inputs.** As described in [adding custom validation](https://www.pantsbuild.org/docs/target-api-new-fields#adding-custom-validation), it's mainly expected that the `Field` will take care of validating itself once you add it. The pants approach is particularly nice because **both the provider and the consumer need to explicitly turn it on**, and because **it attaches a validation process** that is *always* run to check that the output artifacts (libs, headers) actually work. **Spack essentially has all of this, but the validation part isn't shared across packages at all, possibly attracting mild bitrot.**

Separately, this paradigm of *explicit resource dependency on a particular package* and *implicit resource validation* may even be a first step to Spack inferring which resources are available via binary analysis (CMake does a pale but useful imitation of binary analysis [here](https://github.com/Kitware/CMake/blob/v3.19.1/Modules/CheckCXXSourceRuns.cmake)). 

## Proposal

The concrete proposal here is:
1. Separate [all the fine print in our dependency types](https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=dependency%20type#dependency-types) into "resources" which can be depended upon individually.
    - **This "resource dependency" will be implemented using the existing virtual dependencies functionality without modification to any part of the spec syntax.**
    - Create a shared base class/mixin which has hook methods to register resource dependencies with `provides(...)`.
        - **However, `provides()` would be updated to accept a python class as well as a string.**
            - This helps to ensure that spack users and developers are able to see what *type* of dependency it is, and understand very clearly what type of resources are being added to dependee packages.  
    - Each resource class (like the Field concept above) overrides some hook methods to add themselves to the environment correctly.
        - **We have already established this concept with e.g. the `libs()` and `headers()` methods in package.py files -- the only difference we propose is that each of those hook methods be explicitly assigned to a specific purpose-built class definition, instead of requiring it for every package with inheritance.**
        - **Unlike virtual dependencies, resource classes may also perform *validation* upon the result of any of those hook methods.** This may help to fix really annoying bugs.

2. *(Try to)* represent all spack's known deptypes with these resource dependencies, as per https://github.com/spack/spack/issues/3768#issuecomment-293046651.
    - **Once this is done, we would continue to allow `build`, `link`, `run`, and `test` dependencies, but internally they would each be implemented as an umbrella virtual dependency covering each underlying resource dependency for no change in functionality.**

3. *(Try to)* convert all spack *build systems* into collections of resource dependencies that package.py files can choose thoughtfully between.
    - *TODO: this would be a lot better of an argument with more code samples.*

```python
class Resource(ABC):
    pass

@dataclass(frozen=True)
class CFamilyResource(Resource):
    libs: List[Path]
    headers: List[Path]

    # Example of validation.
    def __post_init__(self):
        if not self.headers:
            raise ValueError("bad headers")

...

@dataclass(frozen=True)
class PythonTarget(PythonResource):
    setup_file: Optional[Path]
...

class BuiltWithPythonHasCDependency(Package):
    # The package `some-dep` must have an 
    # explicit `provides(CFamilyResource)` call.
    depends_on('some-dep', CFamilyResource)

    # Possible syntax:
    provides(PythonTarget) # Implies some magic method name.

    # Another possible syntax (like guice):
    @provides(PythonTarget)
    def provides_python_target(self):
        ...
        return PythonTarget(setup_file='setup.py')
```

### Additional information

As described above, pants has used the indirection of `@dataclass` objects and *highly granular dependencies* to avoid surprises in its build logic. This approach is definitely compatible with mypy, for example.

I would appreciate any thoughts on how to make this something spack would actually merge!
